test/harry/main/org/apache/cassandra/harry/model/QuiescentChecker.java

```diff
@@ -1,237 +1,215 @@
-  *
-  *     http://www.apache.org/licenses/LICENSE-2.0
-  *
-  * Unless required by applicable law or agreed to in writing, software
-  * distributed under the License is distributed on an "AS IS" BASIS,
-  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-  * See the License for the specific language governing permissions and
-  * limitations under the License.
-  */
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 
 package org.apache.cassandra.harry.model;
 
-import java.util.*;
-import java.util.function.Supplier;
-
-import javax.annotation.Nullable;
-
-import org.apache.cassandra.harry.core.Run;
-import org.apache.cassandra.harry.data.ResultSetRow;
-import org.apache.cassandra.harry.ddl.ColumnSpec;
-import org.apache.cassandra.harry.ddl.SchemaSpec;
-import org.apache.cassandra.harry.sut.SystemUnderTest;
-import org.apache.cassandra.harry.model.reconciler.PartitionState;
-import org.apache.cassandra.harry.operations.Query;
-import org.apache.cassandra.harry.model.reconciler.Reconciler;
-import org.apache.cassandra.harry.tracker.DataTracker;
-
-import static org.apache.cassandra.harry.gen.DataGenerators.NIL_DESCR;
-import static org.apache.cassandra.harry.gen.DataGenerators.UNSET_DESCR;
-
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NavigableMap;
+
+import accord.utils.Invariants;
+import org.apache.cassandra.harry.execution.DataTracker;
+import org.apache.cassandra.harry.op.Operations;
+import org.apache.cassandra.harry.execution.ResultSetRow;
+import org.apache.cassandra.harry.gen.ValueGenerators;
+
+import static org.apache.cassandra.harry.MagicConstants.LTS_UNKNOWN;
+import static org.apache.cassandra.harry.MagicConstants.NIL_DESCR;
+import static org.apache.cassandra.harry.MagicConstants.NO_TIMESTAMP;
+import static org.apache.cassandra.harry.MagicConstants.UNKNOWN_DESCR;
+import static org.apache.cassandra.harry.MagicConstants.UNSET_DESCR;
+
+/**
+ * Unfortunately model is not reducible to a simple interface of apply/validate, at least not if
+ * we intend to support concurrent validation and in-flight/timed out queries. Validation needs to
```

Comment:
a previous code comment says we don't support timeout (see `org.apache.cassandra.harry.execution.DataTracker`), and here we say we are doing things to support timeout... which is it?  If we see a timeout we fail to update `org.apache.cassandra.harry.execution.DataTracker.SimpleDataTracker#end` so won't be in `finished` but could have been applied, which we will ignore below... so in that case we get an error msg, but its only due to the fact we don't actually support timeout...

The only code I see that also allows concurrent interleaving is `HarrySimualtorTest`, and if this is true then I don't think this logic works either...  lets say you have 3 finished writes and 3 in-flight writes... then the amount of possible states for `List<ResultSetRow> actualRows` is `3!` (`1,2,3 / 1,3,2 / 2,1,3 / 2,3,1 / 3,1,2 / 3,2,1`)... yet this logic expects a single state and validates that the actual row *is* that state...  actually thinking about it more I didn't include the unapplied states... so the possible states is even larger than `3!`...

I feel that *this* class is leaking implementation details throughout the harry source code, making it harder to replace this model with a different model...  stuff like `DataTracker` is leaked and is tightly coupled with this `Model` in question (where a different model may not even use)... this causes the users to copy the state through the code making this code larger and knowing what is allowed vs not harder to understand...

I took the example from `org.apache.cassandra.simulator.paxos.HistoryValidator` as we went through a few iterations (supports concurrent operations and timeout) and came up with this interface

```
public interface Model
{
    Context begin(Visit visit);

    interface Context extends AutoCloseable
    {
        void validate(Operations.SelectStatement select, List<ResultSetRow> actual);

        void writeUnknownFailure();

        @Override
        void close(); // if `writeUnknownFailure` was not called then the mutation is known to have been applied, else its unknown
    }
```

I then did a quick POC and migrated `org.apache.cassandra.distributed.test.log.ResumableStartupTest` to this new interface.  This let me stop propagating `tracker` to multiple places, and really only got used in `org.apache.cassandra.harry.execution.CQLVisitExecutor#execute`... so adding support for failures was not hard

```
public final void execute(Visit visit)
    {
        try (Model.Context context = model.begin(visit))
        {
            CompiledStatement compiledStatement = queryBuilder.compile(visit);
            // All operations are not touching any data
            if (compiledStatement == null)
            {
                Invariants.checkArgument(Arrays.stream(visit.operations).allMatch(op -> op.kind() == Operations.Kind.CUSTOM));
                return;
            }

            List<Operations.SelectStatement> selects = queryBuilder.selects;
            if (selects.isEmpty())
            {
                try
                {
                    executeMutatingVisit(visit, compiledStatement);
                }
                catch (Throwable t)
                {
                    context.writeUnknownFailure();
                }
            }
            else
            {
                Invariants.checkState(selects.size() == 1);
                executeValidatingVisit(context, visit, selects, compiledStatement);
            }
        }
    }

protected void executeValidatingVisit(Model.Context context, Visit visit, List<Operations.SelectStatement> selects, CompiledStatement compiledStatement)
    {
        // TODO: Have not tested with multiple
        List<ResultSetRow> resultSetRow = executeWithResult(visit, compiledStatement);
        try
        {
            context.validate(selects.get(0), resultSetRow);
        }
        catch (Throwable t)
        {
            throw new AssertionError(String.format("Caught an exception while validating %s:\n%s", selects.get(0), compiledStatement),
                                     t);
        }
    }
```

Reply:
for now, concurrent interleaving is not supported.

Reply:
we spoke in slack.  I don't mind leaving this as is for now and sent a POC of migrating to make `DataTracker` a impl detail of the existing checker..

see https://github.com/dcapwell/cassandra/commit/a8a21986a703c42f2cb20b017c0be4cd0b5d934a